lib/stages/BeamBufferSort.hpp

```diff
@@ -0,0 +1,108 @@
+/**
+ * @file
+ * @brief Sort the incoming tracking bream frames by frequency and time and split by frequencies.
+ *  - BeamBufferSort : public kotekan::Stage
+ */
+
+#ifndef BEAM_BUFFER_SORT
+#define BEAM_BUFFER_SORT
+
+#include "BeamMetadata.hpp" // for BeamMetadata
+#include "Config.hpp"       // for Config
+#include "Stage.hpp"
+#include "bufferContainer.hpp" // for bufferContainer
+
+#include <stdint.h> // for uint32_t
+#include <string>   // for string
+#include <vector>   // for vector
+
+
+using std::vector;
+
+
+/**
+ * @class BeamBufferSort
+ * @brief This stage sort the beam buffer by frequency and time and split buffer by frequencies.
+ *
+ * @par Buffers
+ * @buffer in_buf Kotekan single frame tracking beam buffer.
+ *     @buffer_format Array of @c chars
+ * @buffer out_buf Kotekan sorted tracking beam buffer.
+ *     @buffer_format Array of @c uint32_t
+ *
+ * @author Jing Santiago Luo
+ *
+ *
+ */
+
+
+class BeamBufferSort : public kotekan::Stage {
+public:
+    /// Constructor
+    BeamBufferSort(kotekan::Config& config_, const std::string& unique_name,
+                   kotekan::bufferContainer& buffer_container);
+
+
+    /// Destructor
+    virtual ~BeamBufferSort();
+    /// make a frequency ID'ed metadata
+    void make_freq_meta(FreqIDBeamMetadata* out_meta, uint32_t freq_id, 
+        uint64_t fpga_seq_start, timespec ctime, uint64_t stream_id,        
+        dset_id_t dataset_id, uint32_t beam_number, float ra, float dec,
+	uint32_t scaling);
+    /// Transfer a non-frequency id'ed metadata to frequency id'ed metadata
+    void nonfreq_meta_2_freq_meta(BeamMetadata* nonfreq_meta,  
+        FreqIDBeamMetadata* freq_meta, uint32_t freq_id);
+    ///Fill up the empty frame.
+    void fill_empty_frame(uint32_t time_idx, uint32_t freq_idx, 
+        uint64_t fpga_seq_start0, timespec ctime0, uint32_t beam_number, 
+	double ra, double dec, double scaling);
```

Comment:
These should be private, nothing outside of this stage should be calling these functions.